### PR TITLE
feat(list): introduce sort, filtering, pagination models

### PR DESCRIPTION
**Pull request by niyasrad**
## Description

- Added a `meta` object within the `data` object to store pagination details, including the following keys:  
  - `page`: The current page number.  
  - `limit`: The number of items per page.  
  - `total`: The total number of items.

- Introduced a new model for listings, applicable to both config definitions and configs, with the following fields:  
  - `sort_by`: The field to sort by.  
  - `sort_order`: The sorting order (e.g., ascending or descending).  
  - `search`: A search term for filtering results.  
  - `request` (for configs): Used to extract user-applied filters.

- Enhanced validation logic for stricter and faster filtering and sorting.

- Engineered query builders to enable filtering on JSONB models efficiently.

- Added comprehensive test cases to cover all the newly implemented features.  

--- 

## Related Issue

- [x] #41 

## Type of Change

- [ ] Bugfix
- [x] New Feature
- [ ] Documentation

## Checklist

- [x] Code follows the project style guidelines
- [x] Tests have been added/updated
- [x] All tests pass

## Release Notes

- **New Features**
	- Enhanced configuration management API with refined querying capabilities, including sorting and searching options.
	- Added pagination metadata in configuration listing responses.
	- Introduced new functions for dynamic filtering and counting of configuration definitions.
	- Improved handling of request parameters for configuration operations.

- **Bug Fixes**
	- Improved error handling and logging for configuration operations.

- **Tests**
	- Expanded test coverage for sorting, searching, and filtering functionalities.
	- Updated assertions and error messages to reflect changes in parameter terminology.

### CODE REVIEW

File: backend/app/utils/configs/utils.py [hunk 1]
```diff
@@ -128,7 +132,96 @@ def d_config(config_definition_key: str, config_key: str):
...
+
+    -- Returns
+    str
+        The validated or constructed sort field.
+    """
+    if sort_by in allowed_fields:
+        return sort_by
+
+    keys = sort_by.split(".")
+    return l_c_json_path(keys)
```
[the remark this answers is not in the document]
Comment (niyasrad) [reply backend/app/utils/configs/utils.py:170]: Please note that we have the validation done separately in `backend/app/utils/configs/validations.py` - with a method `validate_config_list`. It checks through the indexes, the possible other queryable fields to confirm if the payload is acceptable. There are tests to confirm the same. This applies for any other comment based on "unauthorised fields".

File: backend/app/utils/configs/utils.py [hunk 2]
```diff
@@ -128,7 +132,96 @@ def d_config(config_definition_key: str, config_key: str):
...
+    keys: list[str]
+        The list of keys representing the path in a JSON object.
+
+    -- Returns
+    str
+        The JSON path string.
+    """
+    if len(keys) == 1:
+        return f"data->>'{keys[0]}'"
+    return "data->" + "->".join([f"'{key}'" for key in keys[:-1]]) + f"->>'{keys[-1]}'"
```
[the remark this answers is not in the document]
Comment (niyasrad) [reply backend/app/utils/configs/utils.py:149]: The validations are made early in the API call, and while having edge-cases is important, it is also very important to note that the right modules should be used in the right order, and the functions must be kept as layered as clean as possible. Thank you for the suggestion though.

File: backend/tests/unit_tests/config_definition/payloads/payloads.json [hunk 3]
```diff
@@ -60,127 +134,85 @@
...
+      "params": {
+        "sort_order": "nokey"
+      },
+      "expected_error": "Sort order must be one of 'asc', 'desc'."
+    },
+    "test_list_n_search": {
+      "params": {
+        "search": "123$;"
+      },
+      "expected_error": "Search term must be at least 3 characters long and contain only alphanumeric characters and underscores."
```
[the remark this answers is not in the document]
Comment (niyasrad) [reply backend/tests/unit_tests/config_definition/payloads/payloads.json:215]: I think the current error messages are descriptive and understandable enough. But i'll take this suggestion, and implement it over the time we are changing the error handling.

File: backend/app/api/v1/config_definition.py [hunk 4]
```diff
@@ -135,19 +141,30 @@ def list_config_definition(page: int = 1, limit: int = 10):
...
     """
     try:
-        config_definitions = l_config_definition(page, limit)
+        config_definitions, count = l_config_definition(
+            page, limit, sort_by, sort_order, search
+        )
 
     except Exception as e:
         logger.exception(f"Error listing configuration definitions: {e}")
         raise HTTPException(status_code=500, detail=str(e))
```
[the remark this answers is not in the document]
Comment (niyasrad) [reply backend/app/api/v1/config_definition.py:162]: As discussed, this will be taken up as a part of error handling revamp.

**Review by Vaish24navi — APPROVED** [review]
AMAZING WORK @niyasrad !!!!! 🥇 🥂 🚀

**Pull request merged**